### PR TITLE
1110 ghe : Add BIOS support

### PR DESCRIPTION
**Pull request by cmjishnu**
This commit adds:
1. redfish interface "/Registries/BiosAttributeRegistry" for BIOS Attribute Registry
2. redfish interface "/Systems/system/Bios/Settings" for BIOS settings
3. support for PATCH operation of the bios variables that updates
   the BaseBiosTable.
   
   
 Previous downstream PR : https://github.com/ibm-openbmc/bmcweb/commit/7863e92e690ea4a4752f0bbd22d0cc75f9fbe704

### CODE REVIEW

File: redfish-core/lib/bios.hpp [hunk 1]
```diff
@@ -5,8 +5,141 @@
...
+            std::tuple<std::string, std::variant<int64_t, std::string>>>>>>;
+using BiosBaseTableItemType = std::pair<
+    std::string,
+    std::tuple<
+        std::string, bool, std::string, std::string, std::string,
+        std::variant<int64_t, std::string>, std::variant<int64_t, std::string>,
+        std::vector<
+            std::tuple<std::string, std::variant<int64_t, std::string>>>>>;
+using OptionsItemType =
+    std::tuple<std::string, std::variant<int64_t, std::string>>;
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:33]: Needs

```
#include <array>
#include <string>
#include <string_view>
#include <tuple>
#include <utility>
#include <variant>
#include <vector>
```

File: redfish-core/lib/bios.hpp [hunk 2]
```diff
@@ -43,6 +176,80 @@ inline void
...
                                          true);
+    asyncResp->res.jsonValue["@Redfish.Settings"] = {
+        {"@odata.type", "#Settings.v1_3_0.Settings"},
+        {"SettingsObject",
+         {{"@odata.id", "/redfish/v1/Systems/system/Bios/Settings"}}}};
+    asyncResp->res.jsonValue["AttributeRegistry"] = "BiosAttributeRegistry";
+    asyncResp->res.jsonValue["Attributes"] = {};
+
+    crow::connections::systemBus->async_method_call(
+        [asyncResp](const boost::system::error_code ec,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:187]:
```
const boost::system::error_code& ec
```

File: redfish-core/lib/bios.hpp [hunk 3]
```diff
@@ -43,6 +176,80 @@ inline void
...
+                }
+                else
+                {
+                    BMCWEB_LOG_ERROR("Unsupported attribute type.");
+                    messages::internalError(asyncResp->res);
+                }
+            }
+        },
+            service, "/xyz/openbmc_project/bios_config/manager",
+            "org.freedesktop.DBus.Properties", "Get",
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:245]: Consider to use
```
dbusplus::asio::getProperty<BiosBaseTableType>()
```

File: redfish-core/lib/bios.hpp [hunk 4]
```diff
@@ -43,6 +176,80 @@ inline void
...
+            "xyz.openbmc_project.BIOSConfig.Manager", "BaseBIOSTable");
+    },
+        "xyz.openbmc_project.ObjectMapper",
+        "/xyz/openbmc_project/object_mapper",
+        "xyz.openbmc_project.ObjectMapper", "GetObject",
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:250]: Consider to use 
```
 dbus::utility::getDbusObject()
```

File: redfish-core/lib/bios.hpp [hunk 5]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
 }
 
+/**
+ * BiosSettings class supports handle GET/PATCH method for
+ * BIOS configuration pending settings.
+ */
+inline void requestRoutesBiosSettings(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/Systems/<str>/Bios/Settings")
+        .privileges({{"ConfigureManager"}})
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:270]:
```
        .privileges(redfish::privileges::getBios)
```

File: redfish-core/lib/bios.hpp [hunk 6]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+        },
+            "xyz.openbmc_project.ObjectMapper",
+            "/xyz/openbmc_project/object_mapper",
+            "xyz.openbmc_project.ObjectMapper", "GetObject",
+            "/xyz/openbmc_project/bios_config/manager",
+            std::array<const char*, 0>());
+    });
+
+    BMCWEB_ROUTE(app, "/redfish/v1/Systems/<str>/Bios/Settings")
+        .privileges({{"ConfigureManager"}})
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:367]:
```
        .privileges(redfish::privileges::patchBios)
```

File: redfish-core/lib/bios.hpp [hunk 7]
```diff
@@ -43,6 +176,80 @@ inline void
...
+        {
+            BMCWEB_LOG_ERROR("ObjectMapper::GetObject call failed: {} ", ec);
+            messages::internalError(asyncResp->res);
+
+            return;
+        }
+        const std::string& service = getObjectType.begin()->first;
+
+        crow::connections::systemBus->async_method_call(
+            [asyncResp](const boost::system::error_code ec1,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:199]:
```
const boost::system::error_code& ec1
```

File: redfish-core/lib/bios.hpp [hunk 8]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+            "/redfish/v1/Systems/system/Bios/Settings";
+        asyncResp->res.jsonValue["@odata.type"] = "#Bios.v1_1_0.Bios";
+        asyncResp->res.jsonValue["Name"] = "Bios Settings";
+        asyncResp->res.jsonValue["Id"] = "BiosSettings";
+        asyncResp->res.jsonValue["AttributeRegistry"] = "BiosAttributeRegistry";
+        nlohmann::json attributes(nlohmann::json::value_t::object);
+        asyncResp->res.jsonValue["Attributes"] = attributes;
+
+        crow::connections::systemBus->async_method_call(
+            [asyncResp](const boost::system::error_code ec,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:296]:
```
const boost::system::error_code& ec
```

File: redfish-core/lib/bios.hpp [hunk 9]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+            {
+                BMCWEB_LOG_ERROR("ObjectMapper::GetObject call failed: {}", ec);
+                messages::internalError(asyncResp->res);
+
+                return;
+            }
+            std::string service = getObjectType.begin()->first;
+
+            crow::connections::systemBus->async_method_call(
+                [asyncResp](const boost::system::error_code ec1,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:308]:
```
const boost::system::error_code& ec1
```

File: redfish-core/lib/bios.hpp [hunk 10]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                    }
+                    else
+                    {
+                        BMCWEB_LOG_ERROR("Unsupported attribute type.");
+                        messages::internalError(asyncResp->res);
+                    }
+                }
+            },
+                service, "/xyz/openbmc_project/bios_config/manager",
+                "org.freedesktop.DBus.Properties", "Get",
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:356]: Consider to use
```
getProperty<PendingAttributesType>();
```

File: redfish-core/lib/bios.hpp [hunk 11]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                "xyz.openbmc_project.BIOSConfig.Manager", "PendingAttributes");
+        },
+            "xyz.openbmc_project.ObjectMapper",
+            "/xyz/openbmc_project/object_mapper",
+            "xyz.openbmc_project.ObjectMapper", "GetObject",
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:361]: Consider to use
```
getDbusObject()
```

File: redfish-core/lib/bios.hpp [hunk 12]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+            BMCWEB_LOG_ERROR(
+                "The value for 'Attributes' is in a different format");
+            messages::propertyValueFormatError(asyncResp->res, attrsJson.dump(),
+                                               "Attributes");
+            return;
+        }
+
+        crow::connections::systemBus->async_method_call(
+            [asyncResp, attrsJson,
+             systemName](const boost::system::error_code ec,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:402]:
```
const boost::system::error_code& ec
```

File: redfish-core/lib/bios.hpp [hunk 13]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                else
+                {
+                    BMCWEB_LOG_ERROR("Attribute Type in BiosTable is Unknown");
+                    messages::internalError(asyncResp->res);
+                    return;
+                }
+            }
+
+            crow::connections::systemBus->async_method_call(
+                [asyncResp](const boost::system::error_code ec1) {
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:532]:
```
const boost::system::error_code& ec1
```

File: redfish-core/lib/bios.hpp [hunk 14]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+        asyncResp->res.jsonValue["Name"] = "Bios Attribute Registry";
+        asyncResp->res.jsonValue["Id"] = "BiosAttributeRegistry";
+        asyncResp->res.jsonValue["RegistryVersion"] = "1.0.0";
+        asyncResp->res.jsonValue["Language"] = "en";
+        asyncResp->res.jsonValue["OwningEntity"] = "OpenBMC";
+        asyncResp->res.jsonValue["RegistryEntries"]["Attributes"] =
+            nlohmann::json::array();
+
+        crow::connections::systemBus->async_method_call(
+            [asyncResp](const boost::system::error_code ec,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:584]:
```
const boost::system::error_code& ec
```

File: redfish-core/lib/bios.hpp [hunk 15]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                BMCWEB_LOG_ERROR("ObjectMapper::GetObject call failed: {}", ec);
+                messages::internalError(asyncResp->res);
+
+                return;
+            }
+            std::string service = getObjectType.begin()->first;
+
+            crow::connections::systemBus->async_method_call(
+                [asyncResp](
+                    const boost::system::error_code ec1,
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:597]:
```
const boost::system::error_code& ec1
```

File: redfish-core/lib/bios.hpp [hunk 16]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+
+                    if (!optionsArray.empty())
+                    {
+                        attributeItem["Value"] = optionsArray;
+                    }
+                    attributeArray.push_back(attributeItem);
+                }
+            },
+                service, "/xyz/openbmc_project/bios_config/manager",
+                "org.freedesktop.DBus.Properties", "Get",
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:722]: Consider to use
```
getProperty<BiosBaseTableType>()
```

File: redfish-core/lib/bios.hpp [hunk 17]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                "xyz.openbmc_project.BIOSConfig.Manager", "BaseBIOSTable");
+        },
+            "xyz.openbmc_project.ObjectMapper",
+            "/xyz/openbmc_project/object_mapper",
+            "xyz.openbmc_project.ObjectMapper", "GetObject",
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:727]: Consider to use
```
getDbusObject()
```

**Review by sunharis — APPROVED** [review]

File: redfish-core/lib/bios.hpp [hunk 18]
```diff
@@ -5,8 +5,141 @@
...
+              std::tuple<std::string, std::variant<int64_t, std::string>>>;
+enum PendingAttributesIndex
+{
+    pendingAttrType = 0,
+    pendingAttrValue
+};
+static std::string mapAttrTypeToRedfish(const std::string_view typeDbus)
+{
+    std::string ret;
+    if (typeDbus == "xyz.openbmc_project.BIOSConfig.Manager."
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:70]: New clang-format prefers to keep a string as is.
```
if (typeDbus ==
        "xyz.openbmc_project.BIOSConfig.Manager.AttributeType.Enumeration")
```

File: redfish-core/lib/bios.hpp [hunk 19]
```diff
@@ -5,8 +5,141 @@
...
+                    "AttributeType.Enumeration")
+    {
+        ret = "Enumeration";
+    }
+    else if (typeDbus == "xyz.openbmc_project.BIOSConfig."
+                         "Manager.AttributeType.String")
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:76]: `"xyz.openbmc_project.BIOSConfig.Manager.AttributeType.String"`

File: redfish-core/lib/bios.hpp [hunk 20]
```diff
@@ -5,8 +5,141 @@
...
+    {
+        ret = "String";
+    }
+    else if (typeDbus == "xyz.openbmc_project.BIOSConfig."
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:80]: `"xyz.openbmc_project.BIOSConfig.Manager.AttributeType.Password"`

File: redfish-core/lib/bios.hpp [hunk 21]
```diff
@@ -5,8 +5,141 @@
...
+                         "Manager.AttributeType.Password")
+    {
+        ret = "Password";
+    }
+    else if (typeDbus == "xyz.openbmc_project.BIOSConfig."
+                         "Manager.AttributeType.Integer")
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:86]: `"xyz.openbmc_project.BIOSConfig.Manager.AttributeType.Integer"`

File: redfish-core/lib/bios.hpp [hunk 22]
```diff
@@ -5,8 +5,141 @@
...
+    {
+        ret = "Integer";
+    }
+    else if (typeDbus == "xyz.openbmc_project.BIOSConfig."
+                         "Manager.AttributeType.Boolean")
```
Comment (baemyung) [inline redfish-core/lib/bios.hpp:91]: `"xyz.openbmc_project.BIOSConfig.Manager.AttributeType.Boolean"`

File: redfish-core/lib/bios.hpp (hunk 1 above)
Comment (raviteja-b) [reply redfish-core/lib/bios.hpp:33]: +1

File: redfish-core/lib/bios.hpp (hunk 2 above)
Comment (raviteja-b) [reply redfish-core/lib/bios.hpp:187]: +1

File: redfish-core/lib/bios.hpp [hunk 23]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+            }
+            std::string service = getObjectType.begin()->first;
+
+            crow::connections::systemBus->async_method_call(
+                [asyncResp](const boost::system::error_code ec1,
+                            const std::variant<PendingAttributesType>&
+                                retPendingAttributes) {
+                if (ec1)
+                {
+                    BMCWEB_LOG_ERROR("getBiosSettings DBUS error: {}", ec1);
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:313]: this should be a warning

Comment (gtmills) [reply redfish-core/lib/bios.hpp:313]: if it is an internal error, ERROR is correct here

File: redfish-core/lib/bios.hpp [hunk 24]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+
+        if (!redfish::json_util::readJsonPatch(req, asyncResp->res,
+                                               "Attributes", attrsJson))
+        {
+            return;
+        }
+
+        if (attrsJson.is_array())
+        {
+            BMCWEB_LOG_ERROR(
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:393]: this is a user error so warning here

File: redfish-core/lib/bios.hpp [hunk 25]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+            }
+
+            PendingAttributesType pendingAttributes;
+            for (const auto& attrItr : attrsJson.items())
+            {
+                std::string attrName = attrItr.key();
+
+                if (attrName.empty())
+                {
+                    BMCWEB_LOG_ERROR("Attribute Name cannot be null");
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:440]: user error so warning here

File: redfish-core/lib/bios.hpp [hunk 26]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                if (it == biosAttrsType.end())
+                {
+                    messages::propertyUnknown(asyncResp->res, attrName);
+                    return;
+                }
+
+                bool biosAttrReadOnlyStatus = ((*it).second).first;
+                if (biosAttrReadOnlyStatus)
+                {
+                    BMCWEB_LOG_ERROR(
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:457]: user error so warning here

File: redfish-core/lib/bios.hpp [hunk 27]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                }
+
+                std::string biosRedfishAttrType =
+                    mapAttrTypeToRedfish(biosAttrType);
+                if (biosRedfishAttrType == "Integer")
+                {
+                    if (attrItr.value().type() !=
+                        nlohmann::json::value_t::number_unsigned)
+                    {
+                        BMCWEB_LOG_ERROR("The value must be of type int");
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:478]: user error so warning

File: redfish-core/lib/bios.hpp [hunk 28]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                        attrName, std::make_tuple(biosAttrType, attrValue)));
+                }
+                else if (biosRedfishAttrType == "String" ||
+                         biosRedfishAttrType == "Enumeration" ||
+                         biosRedfishAttrType == "Password")
+                {
+                    if (attrItr.value().type() !=
+                        nlohmann::json::value_t::string)
+                    {
+                        BMCWEB_LOG_ERROR("The value must be of type string");
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:496]: user error so warning

File: redfish-core/lib/bios.hpp [hunk 29]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+                    std::string attrValue = attrItr.value();
+                    pendingAttributes.emplace_back(std::make_pair(
+                        attrName, std::make_tuple(biosAttrType, attrValue)));
+                }
+                else if (biosRedfishAttrType == "Boolean")
+                {
+                    if (attrItr.value().type() !=
+                        nlohmann::json::value_t::boolean)
+                    {
+                        BMCWEB_LOG_ERROR("The value must be of type bool");
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:512]: user error so warning

File: redfish-core/lib/bios.hpp [hunk 30]
```diff
@@ -53,6 +260,476 @@ inline void requestRoutesBiosService(App& app)
...
+            }
+            std::string service = getObjectType.begin()->first;
+
+            crow::connections::systemBus->async_method_call(
+                [asyncResp](
+                    const boost::system::error_code ec1,
+                    const std::variant<BiosBaseTableType>& retBiosTable) {
+                if (ec1)
+                {
+                    BMCWEB_LOG_ERROR("getBiosAttributeRegistry DBUS error: {}",
```
Comment (gtmills) [inline redfish-core/lib/bios.hpp:601]: warning

**Review by gtmills — COMMENTED** [review]
Trying to enforce https://github.com/openbmc/bmcweb/blob/master/DEVELOPING.md#logging-levels like we do upstream so that we can have a discussion later about what logging-level we want

File: redfish-core/lib/bios.hpp (hunk 7 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 9 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 12 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 13 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 14 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 15 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 24 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 25 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 26 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 28 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 22 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 21 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 19 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 18 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

**Comment by baemyung** [comment]
Ci fails  - `clang-format-17` needs to run

File: redfish-core/lib/bios.hpp (hunk 1 above)
Comment (raviteja-b) [reply redfish-core/lib/bios.hpp:33]: @cmjishnu please address this comment

File: redfish-core/lib/bios.hpp (hunk 2 above)
Comment (raviteja-b) [reply redfish-core/lib/bios.hpp:187]: @cmjishnu please address this comment

File: redfish-core/lib/bios.hpp (hunk 23 above)
Comment (raviteja-b) [reply redfish-core/lib/bios.hpp:313]: @cmjishnu modify here to use Warning

File: redfish-core/lib/bios.hpp (hunk 2 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 4 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp:256]: Can I do this in a follow up PR?

File: redfish-core/lib/bios.hpp (hunk 5 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp:276]: Can I do this in a follow up PR?

File: redfish-core/lib/bios.hpp (hunk 23 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp]: Done

File: redfish-core/lib/bios.hpp (hunk 10 above)
Comment (cmjishnu) [reply redfish-core/lib/bios.hpp:364]: Can I do this in a follow up PR?

**Review by baemyung — APPROVED** [review]
Let's merge into 1110, and the other comments can be addressed later.

**Pull request merged**